### ISSUE TITLE
MPI_Abort: Fatal error in MPI_Init: Invalid tag [mvapich2-2.3]

### ISSUE THREAD

**Issue by jameshcorbett** [flux-framework/flux-core#3592]
I built Flux from a development branch that is up-to-date with master, but when I try to run MPI applications with more than one task/rank, I get an "invalid tag" error. Is this expected? (If I remember right, it used to work, but I can't say for sure.) I ran `make check` and all of the tests passed.

Example output:

```
[corbett8@ruby961]$ FLUX_PMI_DEBUG=1 ../flux-core/src/cmd/flux start flux mini run -n2 ./mpi_hello_world
lt-flux-broker: pmi-debug-dlopen: library name /lib64/libpmi.so
pmi-debug-dlopen[-1]: init = operation completed successfully
pmi-debug-dlopen[0]: get_params (rank=0 size=1 kvsname=0.0) = operation completed successfully
pmi-debug-dlopen[0]: kvs_get (kvsname=0.0 key=flux.instance-level value=<none>) = invalid key argument
pmi-debug-dlopen[0]: kvs_get (kvsname=0.0 key=PMI_process_mapping value=<none>) = invalid key argument
pmi-debug-dlopen[0]: finalize = operation completed successfully
0.181s: flux-shell[0]: FATAL: MPI_Abort: Fatal error in MPI_Init: Invalid tag, error stack:
0.181s: flux-shell[0]: stderr: flux-shell: FATAL: MPI_Abort: Fatal error in MPI_Init: Invalid tag, error stack:
0.182s: job.exception type=exec severity=0 MPI_Abort: Fatal error in MPI_Init: Invalid tag, error stack:
flux-job: task(s) exited with exit code 1
2021-04-08T18:42:27.743875Z broker.err[0]: rc2.0: flux mini run -n2 ./mpi_hello_world Exited (rc=1) 0.6s
```

**Comment by garlick** [comment]
The `pmi-debug-delopen` output that you're getting by setting `FLUX_PMI_DEBUG=1` is showing that the broker found SLURM's `libpmi.so` which decided it was not running under slurm, and gave the broker a singleton environment, which is correct.

I think you'll want to enable debugging of the Flux PMI server to see what might be going wrong in your job's `MPI_Init()`.  (Hope that makes sense - this does get confusing).  Here's an example with my "hello" program compiled against openmpi-4.0.3.
```
$ flux start flux mini run -o verbose=2 -n2 t/mpi/hello
0.037s: flux-shell[0]: DEBUG: 0: task_count=2 slot_count=2 cores_per_slot=1 slots_per_node=-1
0.037s: flux-shell[0]: DEBUG: 0: tasks [0-1] on cores 0-1
0.038s: flux-shell[0]: DEBUG: Loading /usr/local/etc/flux/shell/initrc.lua
0.043s: flux-shell[0]: DEBUG: output: batch timeout = 0.500s
0.089s: flux-shell[0]: TRACE: 1: C: cmd=init pmi_version=1 pmi_subversion=1
0.089s: flux-shell[0]: TRACE: 1: S: cmd=response_to_init rc=0 pmi_version=1 pmi_subversion=1
0.089s: flux-shell[0]: TRACE: 1: C: cmd=get_maxes
0.089s: flux-shell[0]: TRACE: 1: S: cmd=maxes rc=0 kvsname_max=64 keylen_max=64 vallen_max=1024
0.090s: flux-shell[0]: TRACE: 1: C: cmd=get_my_kvsname
0.090s: flux-shell[0]: TRACE: 1: S: cmd=my_kvsname rc=0 kvsname=ƒ8eHJG7
0.090s: flux-shell[0]: TRACE: 1: C: cmd=get_my_kvsname
0.090s: flux-shell[0]: TRACE: 1: S: cmd=my_kvsname rc=0 kvsname=ƒ8eHJG7
0.090s: flux-shell[0]: TRACE: 1: C: cmd=get kvsname=ƒ8eHJG7 key=PMI_process_mapping
0.091s: flux-shell[0]: TRACE: 1: S: cmd=get_result rc=4
0.091s: flux-shell[0]: TRACE: 1: C: cmd=get_my_kvsname
0.091s: flux-shell[0]: TRACE: 1: S: cmd=my_kvsname rc=0 kvsname=ƒ8eHJG7
0.091s: flux-shell[0]: TRACE: 1: C: cmd=get kvsname=ƒ8eHJG7 key=PMI_process_mapping
0.091s: flux-shell[0]: TRACE: 1: S: cmd=get_result rc=4
0.091s: flux-shell[0]: TRACE: 1: C: cmd=get_universe_size
0.092s: flux-shell[0]: TRACE: 1: S: cmd=universe_size rc=0 size=2
0.092s: flux-shell[0]: TRACE: 1: C: cmd=get_appnum
0.092s: flux-shell[0]: TRACE: 1: S: cmd=appnum rc=0 appnum=0
0.094s: flux-shell[0]: TRACE: 1: C: cmd=get kvsname=ƒ8eHJG7 key=0-1-key0
0.094s: flux-shell[0]: TRACE: 1: S: cmd=get_result rc=4
0.104s: flux-shell[0]: TRACE: 0: C: cmd=init pmi_version=1 pmi_subversion=1
0.104s: flux-shell[0]: TRACE: 0: S: cmd=response_to_init rc=0 pmi_version=1 pmi_subversion=1
0.104s: flux-shell[0]: TRACE: 0: C: cmd=get_maxes
0.104s: flux-shell[0]: TRACE: 0: S: cmd=maxes rc=0 kvsname_max=64 keylen_max=64 vallen_max=1024
0.104s: flux-shell[0]: TRACE: 0: C: cmd=get_my_kvsname
0.105s: flux-shell[0]: TRACE: 0: S: cmd=my_kvsname rc=0 kvsname=ƒ8eHJG7
0.105s: flux-shell[0]: TRACE: 0: C: cmd=get_my_kvsname
0.105s: flux-shell[0]: TRACE: 0: S: cmd=my_kvsname rc=0 kvsname=ƒ8eHJG7
0.105s: flux-shell[0]: TRACE: 0: C: cmd=get kvsname=ƒ8eHJG7 key=PMI_process_mapping
0.105s: flux-shell[0]: TRACE: 0: S: cmd=get_result rc=4
0.105s: flux-shell[0]: TRACE: 0: C: cmd=get_my_kvsname
0.106s: flux-shell[0]: TRACE: 0: S: cmd=my_kvsname rc=0 kvsname=ƒ8eHJG7
0.106s: flux-shell[0]: TRACE: 0: C: cmd=get kvsname=ƒ8eHJG7 key=PMI_process_mapping
0.106s: flux-shell[0]: TRACE: 0: S: cmd=get_result rc=4
0.106s: flux-shell[0]: TRACE: 0: C: cmd=get_universe_size
0.106s: flux-shell[0]: TRACE: 0: S: cmd=universe_size rc=0 size=2
0.106s: flux-shell[0]: TRACE: 0: C: cmd=get_appnum
0.106s: flux-shell[0]: TRACE: 0: S: cmd=appnum rc=0 appnum=0
0.109s: flux-shell[0]: TRACE: 0: C: cmd=get kvsname=ƒ8eHJG7 key=0-0-key0
0.109s: flux-shell[0]: TRACE: 0: S: cmd=get_result rc=4
0.342s: flux-shell[0]: TRACE: 0: C: cmd=put kvsname=ƒ8eHJG7 key=0-0-key0 value=cG1peC5jcHVzZXQAMDMAMDAwNAAwLTMAcG1peC5obmFtZQAwMwAwMDBkAGppbS1sZW11cnBybwBidGwudGNwLjQuMAAxNAAwMDMwAAoAAQ4AAAAAAAAAAAAAAAAEBAIAAAAAAKwRAAEAAAAAAAAAAAAAAAAEBAMAAAAAAHBtbC5iYXNlLjIuMAAxNAAwMDA0AG9iMQB -
0.343s: flux-shell[0]: TRACE: 0: S: cmd=put_result rc=0
0.343s: flux-shell[0]: TRACE: 0: C: cmd=barrier_in
0.349s: flux-shell[0]: TRACE: 1: C: cmd=put kvsname=ƒ8eHJG7 key=0-1-key0 value=cG1peC5jcHVzZXQAMDMAMDAwNAAwLTMAcG1peC5obmFtZQAwMwAwMDBkAGppbS1sZW11cnBybwBidGwudGNwLjQuMAAxNAAwMDMwAAoAAQ4AAAAAAAAAAAAAAAAEBQIAAAAAAKwRAAEAAAAAAAAAAAAAAAAEBQMAAAAAAD  -
0.349s: flux-shell[0]: TRACE: 1: S: cmd=put_result rc=0
0.350s: flux-shell[0]: TRACE: 1: C: cmd=barrier_in
0.350s: flux-shell[0]: TRACE: 0: S: cmd=barrier_out rc=0
0.350s: flux-shell[0]: TRACE: 1: S: cmd=barrier_out rc=0
0.350s: flux-shell[0]: TRACE: 1: C: cmd=get kvsname=ƒ8eHJG7 key=0-1-key0
0.350s: flux-shell[0]: TRACE: 1: S: cmd=get_result rc=0 value=cG1peC5jcHVzZXQAMDMAMDAwNAAwLTMAcG1peC5obmFtZQAwMwAwMDBkAGppbS1sZW11cnBybwBidGwudGNwLjQuMAAxNAAwMDMwAAoAAQ4AAAAAAAAAAAAAAAAEBQIAAAAAAKwRAAEAAAAAAAAAAAAAAAAEBQMAAAAAAD  -
0.350s: flux-shell[0]: TRACE: 0: C: cmd=barrier_in
0.351s: flux-shell[0]: TRACE: 1: C: cmd=barrier_in
0.351s: flux-shell[0]: TRACE: 0: S: cmd=barrier_out rc=0
0.351s: flux-shell[0]: TRACE: 1: S: cmd=barrier_out rc=0
0.351s: flux-shell[0]: TRACE: 0: C: cmd=get kvsname=ƒ8eHJG7 key=0-1-key0
0.352s: flux-shell[0]: TRACE: 0: S: cmd=get_result rc=0 value=cG1peC5jcHVzZXQAMDMAMDAwNAAwLTMAcG1peC5obmFtZQAwMwAwMDBkAGppbS1sZW11cnBybwBidGwudGNwLjQuMAAxNAAwMDMwAAoAAQ4AAAAAAAAAAAAAAAAEBQIAAAAAAKwRAAEAAAAAAAAAAAAAAAAEBQMAAAAAAD  -
0.352s: flux-shell[0]: TRACE: 1: C: cmd=get kvsname=ƒ8eHJG7 key=0-0-key0
0.352s: flux-shell[0]: TRACE: 1: S: cmd=get_result rc=0 value=cG1peC5jcHVzZXQAMDMAMDAwNAAwLTMAcG1peC5obmFtZQAwMwAwMDBkAGppbS1sZW11cnBybwBidGwudGNwLjQuMAAxNAAwMDMwAAoAAQ4AAAAAAAAAAAAAAAAEBAIAAAAAAKwRAAEAAAAAAAAAAAAAAAAEBAMAAAAAAHBtbC5iYXNlLjIuMAAxNAAwMDA0AG9iMQB -
0: completed MPI_Init in 0.272s.  There are 2 tasks
0: completed first barrier in 0.001s
0: completed MPI_Finalize in 0.004s
0.353s: flux-shell[0]: TRACE: 1: C: cmd=barrier_in
0.353s: flux-shell[0]: TRACE: 0: C: cmd=barrier_in
0.353s: flux-shell[0]: TRACE: 0: S: cmd=barrier_out rc=0
0.353s: flux-shell[0]: TRACE: 1: S: cmd=barrier_out rc=0
0.354s: flux-shell[0]: TRACE: 1: C: cmd=finalize
0.354s: flux-shell[0]: TRACE: 1: S: cmd=finalize_ack rc=0
0.354s: flux-shell[0]: TRACE: 1: S: pmi finalized
0.355s: flux-shell[0]: TRACE: 0: C: cmd=finalize
0.355s: flux-shell[0]: TRACE: 0: S: cmd=finalize_ack rc=0
0.355s: flux-shell[0]: TRACE: 0: S: pmi finalized
0.355s: flux-shell[0]: TRACE: 1: C: pmi EOF
0.355s: flux-shell[0]: TRACE: 0: C: pmi EOF
0.358s: flux-shell[0]: DEBUG: task 1 complete status=0
0.358s: flux-shell[0]: DEBUG: task 0 complete status=0
0.365s: flux-shell[0]: DEBUG: exit 0
```

**Comment by jameshcorbett** [comment]
Thanks---here's what I got:

```
$ ../flux-core/src/cmd/flux start flux mini run -o verbose=2 -n2 ./mpi_hello_world
0.059s: flux-shell[0]: DEBUG: 0: task_count=2 slot_count=2 cores_per_slot=1 slots_per_node=-1
0.059s: flux-shell[0]: DEBUG: 0: tasks [0-1] on cores 0-1
0.061s: flux-shell[0]: DEBUG: Loading /usr/workspace/corbett8/flux-core/src/shell/initrc.lua
0.120s: flux-shell[0]: DEBUG: output: batch timeout = 0.500s
0.308s: flux-shell[0]: TRACE: 1: C: cmd=init pmi_version=1 pmi_subversion=1
0.308s: flux-shell[0]: TRACE: 1: S: cmd=response_to_init rc=0 pmi_version=1 pmi_subversion=1
0.309s: flux-shell[0]: TRACE: 1: C: cmd=get_maxes
0.309s: flux-shell[0]: TRACE: 1: S: cmd=maxes rc=0 kvsname_max=64 keylen_max=64 vallen_max=1024
0.309s: flux-shell[0]: TRACE: 0: C: cmd=init pmi_version=1 pmi_subversion=1
0.309s: flux-shell[0]: TRACE: 0: S: cmd=response_to_init rc=0 pmi_version=1 pmi_subversion=1
0.309s: flux-shell[0]: TRACE: 0: C: cmd=get_maxes
0.309s: flux-shell[0]: TRACE: 0: S: cmd=maxes rc=0 kvsname_max=64 keylen_max=64 vallen_max=1024
0.309s: flux-shell[0]: TRACE: 1: C: cmd=get_appnum
0.309s: flux-shell[0]: TRACE: 1: S: cmd=appnum rc=0 appnum=0
0.309s: flux-shell[0]: TRACE: 0: C: cmd=get_appnum
0.309s: flux-shell[0]: TRACE: 0: S: cmd=appnum rc=0 appnum=0
0.309s: flux-shell[0]: TRACE: 0: C: cmd=get_my_kvsname
0.309s: flux-shell[0]: TRACE: 0: S: cmd=my_kvsname rc=0 kvsname=fr13GE3
0.309s: flux-shell[0]: TRACE: 1: C: cmd=get_my_kvsname
0.309s: flux-shell[0]: TRACE: 1: S: cmd=my_kvsname rc=0 kvsname=fr13GE3
0.309s: flux-shell[0]: TRACE: 0: C: cmd=get kvsname=fr13GE3 key=PMI_process_mapping
0.309s: flux-shell[0]: TRACE: 0: S: cmd=get_result rc=4
0.310s: flux-shell[0]: TRACE: 1: C: cmd=get kvsname=fr13GE3 key=PMI_process_mapping
0.310s: flux-shell[0]: TRACE: 1: S: cmd=get_result rc=4
0.310s: flux-shell[0]: TRACE: 0: C: cmd=abort exit_code=1 error_msg=Fatal error in MPI_Init: Invalid tag, error stack:
0.310s: flux-shell[0]: FATAL: MPI_Abort: Fatal error in MPI_Init: Invalid tag, error stack:
0.310s: flux-shell[0]: stderr: flux-shell: FATAL: pmi/pmi.c:92: MPI_Abort: Fatal error in MPI_Init: Invalid tag, error stack:
0.311s: job.exception type=exec severity=0 MPI_Abort: Fatal error in MPI_Init: Invalid tag, error stack:
flux-job: task(s) exited with exit code 1
2021-04-09T01:52:05.194316Z broker.err[0]: rc2.0: flux mini run -o verbose=2 -n2 ./mpi_hello_world Exited (rc=1) 1.4s
```

**Comment by garlick** [comment]
Hmm, I think that looks OK.  The rc=4 is an "Invalid key" error, but `PMI_process_mapping` _should_ be treated as optional by MPI.  However our code did change recently change in this area and we may be withholding it now in cases where we didn't before - see 7f66c3db865acca1def5ce72681f71b84e1777dc

I wonder if adding `-o pmi.clique=pershell` to the `flux mini run` command would change anything?

Which MPI is this?

(I'm going to be offline for the rest of the day traveling - sorry for that!)

**Comment by jameshcorbett** [comment]
```
[corbett8@quartz770:parsl_tests]$ ldd ./mpi_hello_world | grep mpi
	libmpi.so.12 => /usr/tce/packages/mvapich2/mvapich2-2.3-intel-19.0.4/lib/libmpi.so.12 (0x00002aaaaaccf000)
```

This is just the MPI I got when I used the default `mpicc` command on quartz/ruby. 

It works when I add `-o pmi.clique=pershell`

```
$ ../flux-core/src/cmd/flux start flux mini run -o verbose=2 -o pmi.clique=pershell -n2 ./mpi_hello_world
0.096s: flux-shell[0]: DEBUG: 0: task_count=2 slot_count=2 cores_per_slot=1 slots_per_node=-1
0.096s: flux-shell[0]: DEBUG: 0: tasks [0-1] on cores 0-1
0.097s: flux-shell[0]: DEBUG: Loading /usr/workspace/corbett8/flux-core/src/shell/initrc.lua
0.198s: flux-shell[0]: DEBUG: output: batch timeout = 0.500s
0.431s: flux-shell[0]: TRACE: 0: C: cmd=init pmi_version=1 pmi_subversion=1
0.431s: flux-shell[0]: TRACE: 0: S: cmd=response_to_init rc=0 pmi_version=1 pmi_subversion=1
0.431s: flux-shell[0]: TRACE: 1: C: cmd=init pmi_version=1 pmi_subversion=1
0.431s: flux-shell[0]: TRACE: 1: S: cmd=response_to_init rc=0 pmi_version=1 pmi_subversion=1
0.431s: flux-shell[0]: TRACE: 0: C: cmd=get_maxes
0.432s: flux-shell[0]: TRACE: 0: S: cmd=maxes rc=0 kvsname_max=64 keylen_max=64 vallen_max=1024
0.432s: flux-shell[0]: TRACE: 1: C: cmd=get_maxes
0.432s: flux-shell[0]: TRACE: 1: S: cmd=maxes rc=0 kvsname_max=64 keylen_max=64 vallen_max=1024
0.432s: flux-shell[0]: TRACE: 1: C: cmd=get_appnum
0.432s: flux-shell[0]: TRACE: 1: S: cmd=appnum rc=0 appnum=0
0.432s: flux-shell[0]: TRACE: 0: C: cmd=get_appnum
0.432s: flux-shell[0]: TRACE: 0: S: cmd=appnum rc=0 appnum=0
0.432s: flux-shell[0]: TRACE: 1: C: cmd=get_my_kvsname
0.432s: flux-shell[0]: TRACE: 1: S: cmd=my_kvsname rc=0 kvsname=f2QJycRu
0.432s: flux-shell[0]: TRACE: 0: C: cmd=get_my_kvsname
0.432s: flux-shell[0]: TRACE: 0: S: cmd=my_kvsname rc=0 kvsname=f2QJycRu
0.432s: flux-shell[0]: TRACE: 1: C: cmd=get kvsname=f2QJycRu key=PMI_process_mapping
0.432s: flux-shell[0]: TRACE: 1: S: cmd=get_result rc=0 value=(vector,(0,1,2))
0.432s: flux-shell[0]: TRACE: 0: C: cmd=get kvsname=f2QJycRu key=PMI_process_mapping
0.432s: flux-shell[0]: TRACE: 0: S: cmd=get_result rc=0 value=(vector,(0,1,2))
0.510s: flux-shell[0]: TRACE: 1: C: cmd=put kvsname=f2QJycRu key=pmi_ahkey_1 value=0000001900000016
0.511s: flux-shell[0]: TRACE: 1: S: cmd=put_result rc=0
0.511s: flux-shell[0]: TRACE: 0: C: cmd=put kvsname=f2QJycRu key=pmi_ahkey_0 value=0000001900000016
0.511s: flux-shell[0]: TRACE: 0: S: cmd=put_result rc=0
0.511s: flux-shell[0]: TRACE: 1: C: cmd=barrier_in
0.511s: flux-shell[0]: TRACE: 0: C: cmd=barrier_in
0.511s: flux-shell[0]: TRACE: 0: S: cmd=barrier_out rc=0
0.511s: flux-shell[0]: TRACE: 1: S: cmd=barrier_out rc=0
0.511s: flux-shell[0]: TRACE: 1: C: cmd=get kvsname=f2QJycRu key=pmi_ahkey_0
0.511s: flux-shell[0]: TRACE: 1: S: cmd=get_result rc=0 value=0000001900000016
0.511s: flux-shell[0]: TRACE: 0: C: cmd=get kvsname=f2QJycRu key=pmi_ahkey_1
0.511s: flux-shell[0]: TRACE: 0: S: cmd=get_result rc=0 value=0000001900000016
0.512s: flux-shell[0]: TRACE: 1: C: cmd=barrier_in
0.514s: flux-shell[0]: TRACE: 0: C: cmd=barrier_in
0.514s: flux-shell[0]: TRACE: 0: S: cmd=barrier_out rc=0
0.514s: flux-shell[0]: TRACE: 1: S: cmd=barrier_out rc=0
0.520s: flux-shell[0]: TRACE: 1: C: cmd=put kvsname=f2QJycRu key=pmi_epidkey_1 value=182450210734090
0.520s: flux-shell[0]: TRACE: 1: S: cmd=put_result rc=0
0.520s: flux-shell[0]: TRACE: 0: C: cmd=put kvsname=f2QJycRu key=pmi_epidkey_0 value=182445915766794
0.520s: flux-shell[0]: TRACE: 0: S: cmd=put_result rc=0
0.520s: flux-shell[0]: TRACE: 1: C: cmd=barrier_in
0.520s: flux-shell[0]: TRACE: 0: C: cmd=barrier_in
0.520s: flux-shell[0]: TRACE: 0: S: cmd=barrier_out rc=0
0.520s: flux-shell[0]: TRACE: 1: S: cmd=barrier_out rc=0
0.520s: flux-shell[0]: TRACE: 0: C: cmd=get kvsname=f2QJycRu key=pmi_epidkey_0
0.520s: flux-shell[0]: TRACE: 0: S: cmd=get_result rc=0 value=182445915766794
0.520s: flux-shell[0]: TRACE: 1: C: cmd=get kvsname=f2QJycRu key=pmi_epidkey_0
0.520s: flux-shell[0]: TRACE: 1: S: cmd=get_result rc=0 value=182445915766794
0.521s: flux-shell[0]: TRACE: 1: C: cmd=get kvsname=f2QJycRu key=pmi_epidkey_1
0.521s: flux-shell[0]: TRACE: 1: S: cmd=get_result rc=0 value=182450210734090
0.521s: flux-shell[0]: TRACE: 0: C: cmd=get kvsname=f2QJycRu key=pmi_epidkey_1
0.521s: flux-shell[0]: TRACE: 0: S: cmd=get_result rc=0 value=182450210734090
0.527s: flux-shell[0]: TRACE: 0: C: cmd=finalize
0.527s: flux-shell[0]: TRACE: 0: S: cmd=finalize_ack rc=0
0.527s: flux-shell[0]: TRACE: 0: S: pmi finalized
0.527s: flux-shell[0]: TRACE: 0: C: pmi EOF
0.527s: flux-shell[0]: TRACE: 1: C: cmd=finalize
0.527s: flux-shell[0]: TRACE: 1: S: cmd=finalize_ack rc=0
0.527s: flux-shell[0]: TRACE: 1: S: pmi finalized
0.528s: flux-shell[0]: TRACE: 1: C: pmi EOF
Hello world from processor quartz770, rank 1 out of 2 processors
Hello world from processor quartz770, rank 0 out of 2 processors
0.535s: flux-shell[0]: DEBUG: task 1 complete status=0
0.536s: flux-shell[0]: DEBUG: task 0 complete status=0
0.547s: flux-shell[0]: DEBUG: exit 0
```

**Comment by garlick** [comment]
Thanks @jameshcorbett - that is helpful.

Hrm.  So we got to this behavior to resolve an openmpi segfault (#3551) and now mvapich is segfaulting.  I guess the challenge is to figure out how to make both happy automatically.  Also, clearly we would benefit from having a matrix of MPI tests for flux.

**Comment by garlick** [comment]
Reopening as a user on quartz is hitting this again.  Reproducer using `flux-core-0.26.0-140-g2bef19bc1` (thanks @dongahn) slightly modified by me:
```
$ cd flux-core
$ srun -ppdebug --pty --mpi=none --mpibind=off -N2 src/cmd/flux start \
    flux mini run -n2 -N2 t/mpi/hello
```
mvpapich2 is failing as above, e.g. get of `PMI_process_mapping`is failing with Invalid key.

The root cause is that slurm does not provide `PMI_process_mapping` to flux when `--mpi=none` is set.   This means the `broker.mapping` attribute will not be set and the shell PMI plugin will not be able to determine the mapping for flux jobs.

Since mvapich2 has (uniquely?) decided that it must have `PMI_process_mapping` set, perhaps the solution is to work harder to create it when it is not provided to us by enclosing RM (slurm in this case).

**Comment by garlick** [comment]
The rationale for PR #3553 which fixed openmpi, but broke mvapich2 in 0.25.0 is summarized in this comment:
https://github.com/flux-framework/flux-core/issues/3551#issuecomment-796929839

Possibly it will no longer be a problem to use "pershell" as the fallback mapping after #3661 is merged.  That would make mvapich.  I will see if I can test that combination.
